addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java

```diff
@@ -0,0 +1,208 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CommandClass getCommandClass() {
+        return CommandClass.TIME_PARAMETERS;
+    }
+
+    /**
+     * Gets a SerialMessage with the TIME_GET command
+     *
+     * @return the serial message.
+     */
+    @Override
+    public SerialMessage getValueMessage() {
+        logger.debug("NODE {}: Creating new message for command TIME_GET", getNode().getNodeId());
+
+        SerialMessage result = new SerialMessage(getNode().getNodeId(), SerialMessageClass.SendData,
+                SerialMessageType.Request, SerialMessageClass.ApplicationCommandHandler, SerialMessagePriority.Get);
+        ByteArrayOutputStream outputData = new ByteArrayOutputStream();
+        outputData.write(getNode().getNodeId());
+        outputData.write(2);
+        outputData.write(getCommandClass().getKey());
+        outputData.write(TIME_GET);
+        result.setMessagePayload(outputData.toByteArray());
+        return result;
+    }
+
+    /**
+     * Gets a SerialMessage with the TIME_SET command
+     *
+     * @return the serial message.
+     */
+    public SerialMessage getSetMessage(Date date) {
+        logger.debug("NODE {}: Creating new message for command TIME_SET", getNode().getNodeId());
+
+        Calendar cal = Calendar.getInstance();
+        cal.setTime(date);
+
+        SerialMessage result = new SerialMessage(getNode().getNodeId(), SerialMessageClass.SendData,
+                SerialMessageType.Request, SerialMessageClass.SendData, SerialMessagePriority.Immediate);
+
+        ByteArrayOutputStream outputData = new ByteArrayOutputStream();
+        outputData.write(getNode().getNodeId());
+        outputData.write(9);
+        outputData.write(getCommandClass().getKey());
+        outputData.write(TIME_SET);
+        outputData.write((cal.get(Calendar.YEAR) & 0xff00) >> 8);
+        outputData.write((cal.get(Calendar.YEAR) & 0xff));
+        outputData.write(cal.get(Calendar.MONTH) + 1);
+        outputData.write(cal.get(Calendar.DAY_OF_MONTH));
+        outputData.write(cal.get(Calendar.HOUR_OF_DAY));
+        outputData.write(cal.get(Calendar.MINUTE));
+        outputData.write(cal.get(Calendar.SECOND));
+        result.setMessagePayload(outputData.toByteArray());
+        return result;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @throws ZWaveSerialMessageException
+     */
+    @Override
+    public void handleApplicationCommandRequest(SerialMessage serialMessage, int offset, int endpoint)
+            throws ZWaveSerialMessageException {
+        logger.debug("NODE {}: Received clock command (v{})", this.getNode().getNodeId(), this.getVersion());
+        int command = serialMessage.getMessagePayloadByte(offset);
+        switch (command) {
+            case TIME_GET:
+                logger.debug("NODE %d: requested the time", getNode().getNodeId());
+                getController().sendData(getSetMessage(new Date()));
```

Comment:
I'm not a fan (in general) of calling back to the controller. I know it's done quite a lot in the binding (at the moment), but I'd quite like to reduce or eliminate it where possible...

Additionally, I think we should move all the 'time set' functionallity out of the command classes and into the application where there is more flexibility and knowledge of the environment.